File: api-cmssignature/csharp/src/signature/ESX1.cs
using System;
using System.Collections.Generic;
using tr.gov.tubitak.uekae.esya.api.asn.cms;
using tr.gov.tubitak.uekae.esya.api.cmssignature.attribute;
using tr.gov.tubitak.uekae.esya.api.cmssignature.bundle;
using tr.gov.tubitak.uekae.esya.api.cmssignature.validation.check;

namespace tr.gov.tubitak.uekae.esya.api.cmssignature.signature
{
    public class ESX1 : ESC
    {
        public ESX1(BaseSignedData aSD)
            : base(aSD)
        {
            mSignatureType = ESignatureType.TYPE_ESX_Type1;
        }

        public ESX1(BaseSignedData aSD, ESignerInfo aSigner)
            : base(aSD, aSigner)
        {

            mSignatureType = ESignatureType.TYPE_ESX_Type1;
        }

        //@Override
        protected override void _addUnsignedAttributes(Dictionary<String, Object> aParameters)
        {
            base._addUnsignedAttributes(aParameters);
            _addESX1Attributes(aParameters);
        }

        private void _addESX1Attributes(Dictionary<String, Object> aParameters)
        {
            _addTSCertRevocationValues(aParameters, AttributeOIDs.id_aa_signatureTimeStampToken, true);

            CAdES_C_TimeStampAttr timestamp = new CAdES_C_TimeStampAttr();
            timestamp.setParameters(aParameters);
            //aParameters.put(EParameter.P_SIGNER_INFO, aSignerInfo);
            timestamp.setValue();
            mSignerInfo.addUnsignedAttribute(timestamp.getAttribute());

            Boolean validateTS = (Boolean) aParameters[EParameters.P_VALIDATE_TIMESTAMP_WHILE_SIGNING];

		if (validateTS){
			CAdES_C_TimeStampAttrChecker cAdES_C_TimeStampAttrChecker=new CAdES_C_TimeStampAttrChecker();
			Dictionary<String,Object> paramaters = new Dictionary<String, Object>(aParameters);
		    paramaters.Remove(AllEParameters.P_SIGNING_CERTIFICATE);
            cAdES_C_TimeStampAttrChecker.setParameters(paramaters);
			CheckerResult aCheckerResult =new CheckerResult();
			Boolean result=cAdES_C_TimeStampAttrChecker.check(this, aCheckerResult);
			if(!result){
				logger.Error(aCheckerResult.ToString());
                throw new CMSSignatureException(Msg.getMsg(Msg.CADES_C_TIMESTAMP_INVALID));
			}
		}
        }

        //@Override
        override /*protected*/internal void _convert(ESignatureType aType, Dictionary<String, Object> aParameters)
        {
            aParameters[AllEParameters.P_SIGNER_INFO] = mSignerInfo;
            if (aType.Equals(ESignatureType.TYPE_BES) || aType.Equals(ESignatureType.TYPE_EPES) || aType == ESignatureType.TYPE_EST)
            {
                base._convert(aType, aParameters);
                _addESX1Attributes(aParameters);
            }
            else if (aType == ESignatureType.TYPE_ESC || aType == ESignatureType.TYPE_ESX_Type1)
            {
                _addESX1Attributes(aParameters);
            }
            else
                throw new CMSSignatureException("Signature type:" + aType.name() + " can not be converted to ES_X1");

        }

        /*@Override
        public ESignatureType getType() 
        {
            return ESignatureType.TYPE_ESX_Type1;
        }*/
    }
}

File: api-cmssignature/src/tr/gov/tubitak/uekae/esya/api/cmssignature/signature/ESX1.java
package tr.gov.tubitak.uekae.esya.api.cmssignature.signature;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import tr.gov.tubitak.uekae.esya.api.asn.cms.EAttribute;
import tr.gov.tubitak.uekae.esya.api.asn.cms.ESignerInfo;
import tr.gov.tubitak.uekae.esya.api.cmssignature.CMSSignatureException;
import tr.gov.tubitak.uekae.esya.api.cmssignature.CertRevocationInfoFinder.CertRevocationInfo;
import tr.gov.tubitak.uekae.esya.api.cmssignature.attribute.AttributeOIDs;
import tr.gov.tubitak.uekae.esya.api.cmssignature.attribute.CAdES_C_TimeStampAttr;
import tr.gov.tubitak.uekae.esya.api.cmssignature.attribute.AllEParameters;
import tr.gov.tubitak.uekae.esya.api.cmssignature.attribute.EParameters;
import tr.gov.tubitak.uekae.esya.api.cmssignature.bundle.CMSSignatureI18n;
import tr.gov.tubitak.uekae.esya.api.cmssignature.bundle.E_KEYS;
import tr.gov.tubitak.uekae.esya.api.cmssignature.validation.check.CAdES_C_TimeStampAttrChecker;
import tr.gov.tubitak.uekae.esya.api.cmssignature.validation.check.CheckerResult;

public class ESX1 extends ESC {
	ESX1(BaseSignedData aSD)
	{
		super(aSD);
		mSignatureType = ESignatureType.TYPE_ESX_Type1;
	}
	
	ESX1(BaseSignedData aSD,ESignerInfo aSigner)
	{
		super(aSD,aSigner);
		mSignatureType = ESignatureType.TYPE_ESX_Type1;
	}
	
	@Override
	protected void _addUnsignedAttributes(Map<String, Object> aParameters) 
	throws CMSSignatureException 
	{
		super._addUnsignedAttributes(aParameters);
		_addESX1Attributes(aParameters);
	}
	
	private void _addESX1Attributes(Map<String, Object> aParameters) 
	throws CMSSignatureException
	{
		CAdES_C_TimeStampAttr timestamp = new CAdES_C_TimeStampAttr();
		timestamp.setParameters(aParameters);

        _addTSCertRevocationValues(aParameters, AttributeOIDs.id_aa_signatureTimeStampToken, true);
		
		//aParameters.put(EParameter.P_SIGNER_INFO, aSignerInfo);
		timestamp.setValue();
		mSignerInfo.addUnsignedAttribute(timestamp.getAttribute());
		
		Boolean validateTS = (Boolean) aParameters.get(EParameters.P_VALIDATE_TIMESTAMP_WHILE_SIGNING);

		if (validateTS){
			CAdES_C_TimeStampAttrChecker cAdES_C_TimeStampAttrChecker=new CAdES_C_TimeStampAttrChecker();
			Map<String,Object> params = new HashMap<String, Object>(aParameters);
		    params.remove(AllEParameters.P_SIGNING_CERTIFICATE);
		    cAdES_C_TimeStampAttrChecker.setParameters(params);
			CheckerResult aCheckerResult =new CheckerResult();
			boolean result=cAdES_C_TimeStampAttrChecker.check(this, aCheckerResult);
			if(!result){
				logger.error(aCheckerResult.toString());
				throw new CMSSignatureException(CMSSignatureI18n.getMsg(E_KEYS.CADES_C_TIMESTAMP_INVALID));
			}
		}
	
	}
	
	@Override
	protected void _convert(ESignatureType aType,Map<String, Object> aParameters) throws CMSSignatureException {
		
		aParameters.put(AllEParameters.P_SIGNER_INFO, mSignerInfo);
		if(aType.equals(ESignatureType.TYPE_BES) ||aType.equals(ESignatureType.TYPE_EPES) || aType == ESignatureType.TYPE_EST)
		{
			super._convert(aType, aParameters);
			_addESX1Attributes(aParameters);
		}
		else if(aType == ESignatureType.TYPE_ESC || aType == ESignatureType.TYPE_ESX_Type1 )
		{
			
			_addESX1Attributes(aParameters);
		}
		else
			throw new CMSSignatureException("Signature type:"+aType.name()+" can not be converted to ES_X1");
		
	}
	
	/*@Override
	public ESignatureType getType() 
	{
		return ESignatureType.TYPE_ESX_Type1;
	}*/
}
